examples/common/imgui_ui/ui.cpp

```diff
@@ -0,0 +1,219 @@
+    SDL_GL_SetSwapInterval(1); // Enable vsync
+
+    // Setup Dear ImGui context
+    IMGUI_CHECKVERSION();
+    ImGui::CreateContext();
+    ImGuiIO & io = ImGui::GetIO();
+    (void) io;
+    // io.ConfigFlags |= ImGuiConfigFlags_NavEnableKeyboard;     // Enable Keyboard Controls
+    // io.ConfigFlags |= ImGuiConfigFlags_NavEnableGamepad;      // Enable Gamepad Controls
+
+    ImGui::StyleColorsDark();
+
+    // Setup Platform/Renderer backends
+    ImGui_ImplSDL2_InitForOpenGL(*window, *gl_context);
+    ImGui_ImplOpenGL3_Init();
+}
+
+void UiShutdown(SDL_GLContext * gl_context, SDL_Window ** window)
+{
+    ImGui_ImplOpenGL3_Shutdown();
+    ImGui_ImplSDL2_Shutdown();
+    ImGui::DestroyContext();
+
+    SDL_GL_DeleteContext(*gl_context);
+    SDL_DestroyWindow(*window);
+    SDL_Quit();
+}
+
+void EventLoop(ImguiUi * ui)
+{
+    gUiRunning = true;
+    SDL_GLContext gl_context;
+    SDL_Window * window = nullptr;
+
+    UiInit(&gl_context, &window);
+
+    ImGuiIO & io = ImGui::GetIO();
+
+    while (gUiRunning.load())
+    {
+        SDL_Event event;
+        while (SDL_PollEvent(&event))
+        {
+            ImGui_ImplSDL2_ProcessEvent(&event);
+            if ((event.type == SDL_QUIT) ||
+                (event.type == SDL_WINDOWEVENT && event.window.event == SDL_WINDOWEVENT_CLOSE &&
+                 event.window.windowID == SDL_GetWindowID(window)))
+            {
+                gUiRunning = false;
+            }
+        }
+
+        ImGui_ImplOpenGL3_NewFrame();
+        ImGui_ImplSDL2_NewFrame();
+        ImGui::NewFrame();
+
+        ui->UpdateState();
+        ui->Render();
+
+        // rendering
+        ImGui::Render();
+        glViewport(0, 0, (int) io.DisplaySize.x, (int) io.DisplaySize.y);
+        glClearColor(0, 0, 0, 0);
+        glClear(GL_COLOR_BUFFER_BIT);
+        ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
+        SDL_GL_SwapWindow(window);
+    }
+
+    UiShutdown(&gl_context, &window);
+
+    ChipLogProgress(AppServer, "UI thread Stopped...");
+}
+
+} // namespace
+
+void ImguiUi::RunMainLoop()
+{
+    // Guaranteed to be on the main task (no chip event loop started yet)
+    ChipLoopLoadInitialState();
+
+    // Platform event loop will be on a separate thread,
+    // while the event UI loop will be on the main thread.
+    chip::DeviceLayer::PlatformMgr().StartEventLoopTask();
+
+    // SignalSafeStopMainLoop will stop this loop below
+    // or the loop exits by itself when processing a SDL
+    // exit (generally by clicking the window close icon).
+    EventLoop(this);
+
+    // ensure shutdown events are generated (generally basic cluster
+    // will send a shutdown event to subscribers)
+    chip::DeviceLayer::PlatformMgr().ScheduleWork([](intptr_t) { chip::DeviceLayer::PlatformMgr().HandleServerShuttingDown(); });
+
+    // Stop the chip main loop as well. This is expected to
+    // wait for the task to finish.
+    chip::DeviceLayer::PlatformMgr().StopEventLoopTask();
```

Comment:
This does not guarantee that the thing we just scheduled will run.  See API documentation....  So we might in fact not generate the shutdown event.  In fact, unless we lose the timeslice between line 159 and 163 there is a _very_ good chance that we will not generate it.

That's fine, since the event generation is optional, but the comments should reflect reality.

If we want to wait on the event finishing, what we probably want to do here is to block until the event unblocks us (e.g. via another semaphore).

Reply:
There's no MRP for shutdown events anyway, since we stop immediately after.

To be clear, I am OK with saying that we just might not generate them.  But the comment about "expected to wait for the task to finish" makes it sound like we will wait for the _queued_ task, and that's just not true.